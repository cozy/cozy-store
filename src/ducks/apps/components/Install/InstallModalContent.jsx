import React, { Component } from 'react'

import { ModalContent } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'

import { translate } from 'cozy-ui/react/I18n'
import { SubTitle } from 'cozy-ui/react/Text'
import AnimatedModalHeader from 'ducks/components/AnimatedModalHeader'
import TransparencyLabel from '../TransparencyLabel'

export class InstallModalContent extends Component {
  render() {
    const { app, fetchError, isFetching, isCanceling, t } = this.props
    const isFirstLoading = isFetching && !isCanceling
    // this part must not be wrapped in a component
    // so we get the content using it as a function
    const animatedHeader = AnimatedModalHeader({
      app
    })

    return isFirstLoading ? (
      <ModalContent className="sto-modal-content">
        <div className="sto-install-loading">
          <Spinner size="xlarge" />
        </div>
      </ModalContent>
    ) : (
      <ModalContent className="sto-modal-content">
        {animatedHeader}
        <SubTitle className="sto-modal-title">
          {t('app_modal.install.title')}
        </SubTitle>
        <TransparencyLabel app={app} />
        {fetchError && (
          <p className="u-error">
            {t('app_modal.install.message.version_error', {
              message: fetchError.message
            })}
          </p>
        )}
      </ModalContent>
    )
  }
}

export default translate()(InstallModalContent)
