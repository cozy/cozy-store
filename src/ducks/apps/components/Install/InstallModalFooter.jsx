import React, { Component } from 'react'

import { ModalFooter } from 'cozy-ui/react/Modal'

import { translate } from 'cozy-ui/react/I18n'
import Icon from 'cozy-ui/react/Icon'

import cozySmileIcon from 'assets/icons/icon-cozy-smile.svg'

export class InstallModalFooter extends Component {
  installApp() {
    this.setState({ error: null })
    const { app, onError, channel } = this.props
    this.props.installApp(app.slug, app.type, channel).catch(error => {
      if (onError) return onError(error)
      throw error
    })
  }

  render() {
    const {
      fetchError,
      installError,
      isFetching,
      isInstalling,
      isCanceling,
      t
    } = this.props

    const isFirstLoading = isFetching && !isCanceling
    return (
      !isFirstLoading &&
      !fetchError && (
        <ModalFooter>
          {installError && (
            <p className="u-error">
              {t('app_modal.install.message.install_error', {
                message: installError.message
              })}
            </p>
          )}
          <div className="sto-install-controls">
            <button
              role="button"
              disabled={isInstalling || isCanceling}
              aria-busy={isInstalling}
              className="c-btn c-btn--regular c-btn--download"
              onClick={() => this.installApp()}
            >
              <Icon
                icon={cozySmileIcon}
                color="#FFFFFF"
                width="16px"
                height="16px"
                className="sto-app-install-button-icon"
              />
              <span>{t('app_modal.install.install')}</span>
            </button>
          </div>
        </ModalFooter>
      )
    )
  }
}

export default translate()(InstallModalFooter)
