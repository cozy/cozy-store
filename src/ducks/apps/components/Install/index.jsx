import React, { Component } from 'react'
import { withRouter } from 'react-router'

import Modal from 'cozy-ui/react/Modal'

import getChannel from 'lib/getChannelFromSource'

import InstallModalContent from './InstallModalContent'
import InstallModalFooter from './InstallModalFooter'

export class Install extends Component {
  constructor(props) {
    super(props)
    this.gotoParent = this.gotoParent.bind(this)
    this.state = {
      previousChannel: props.channel ? getChannel(props.app.source) : null,
      isCanceling: false
    }
    if (typeof props.fetchApp === 'function') props.fetchApp(props.channel)
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.app) this.gotoParent()
  }

  async gotoParent() {
    const { app, parent, history, fetchApp } = this.props
    const { previousChannel } = this.state
    // fetch previous channel if channel switch canceled
    if (previousChannel && typeof fetchApp === 'function') {
      this.setState(() => ({ isCanceling: true }))
      await fetchApp(previousChannel)
      this.setState(() => ({ isCanceling: false }))
    }

    if (app && app.slug) {
      history.push(`${parent}/${app.slug}`)
    } else {
      history.push(parent)
    }
  }

  render() {
    const { app, installApp, isInstalling, channel, isAppFetching } = this.props
    const { isCanceling } = this.state
    if (!app) return null
    return (
      <div className="sto-modal--install">
        <Modal dismissAction={this.gotoParent} mobileFullscreen>
          <InstallModalContent
            app={app}
            isFetching={isAppFetching}
            isCanceling={isCanceling}
          />
          <InstallModalFooter
            app={app}
            installApp={installApp}
            isFetching={isAppFetching}
            channel={channel}
            isInstalling={isInstalling}
            isCanceling={isCanceling}
          />
        </Modal>
      </div>
    )
  }
}

export default withRouter(Install)
