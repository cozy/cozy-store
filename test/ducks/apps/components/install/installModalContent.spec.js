'use strict'

/* eslint-env jest */

import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { tMock } from '../../../../jestLib/I18n'
import { InstallModalContent } from 'ducks/apps/components/Install/InstallModalContent'
import mockApp from '../../_mockPhotosRegistryVersion'

const mockFetchError = new Error('Fetch mock error')

Enzyme.configure({ adapter: new Adapter() })

const getProps = () => ({
  t: tMock,
  app: mockApp
})

describe('InstallModalContent modal component', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <InstallModalContent {...getProps()} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should return a spinner if isFetching', () => {
    const component = shallow(
      <InstallModalContent {...getProps()} isFetching />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should have install button busy if isInstalling', () => {
    const component = shallow(
      <InstallModalContent {...getProps()} isInstalling />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should not display spinner if isFetching but during a cancel (isCanceling)', () => {
    const component = shallow(
      <InstallModalContent {...getProps()} isFetching isCanceling />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display error if fetchError', () => {
    const component = shallow(
      <InstallModalContent {...getProps()} fetchError={mockFetchError} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
