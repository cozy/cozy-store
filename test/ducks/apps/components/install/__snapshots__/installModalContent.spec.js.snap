// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallModalContent modal component should be rendered correctly 1`] = `
<ModalContent
  className="sto-modal-content"
>
  <AnimatedContentHeader
    activeClassName="sto-animated-header--active"
    className="sto-animated-header"
  >
    <Icon
      className="sto-animated-header-icon blurry"
      color="#95999D"
      height="64"
      icon="test-file-stub"
      spin={false}
      width="48"
    />
    <Icon
      className="sto-animated-header-exchange"
      color="#95999D"
      icon="exchange"
      spin={false}
    />
    <Icon
      className="sto-animated-header-icon"
      height="56"
      icon="test-file-stub"
      spin={false}
      width="56"
    />
  </AnimatedContentHeader>
  <Unknown
    className="sto-modal-title"
    ellipsis={false}
    tag="div"
  >
    Data access requirements
  </Unknown>
  <withRouter(Wrapper)
    app={
      Object {
        "editor": "Cozy",
        "manifest": Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/3.0.0/icon.svg",
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "permissions": Object {
            "albums": Object {
              "description": "Required to manage photos albums",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.photos.albums",
            },
            "apps": Object {
              "description": "Required by the cozy-bar to display the icons of the apps",
              "type": "io.cozy.apps",
              "verbs": Array [
                "GET",
              ],
            },
            "contacts": Object {
              "description": "Required to to share photos with your contacts",
              "methods": Array [
                "GET",
                "POST",
              ],
              "type": "io.cozy.contacts",
            },
            "files": Object {
              "description": "Required for photo access",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.files",
            },
            "settings": Object {
              "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
              "type": "io.cozy.settings",
              "verbs": Array [
                "GET",
              ],
            },
          },
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        },
        "slug": "photos",
        "type": "webapp",
        "version": "3.0.0",
      }
    }
  />
</ModalContent>
`;

exports[`InstallModalContent modal component should display error if fetchError 1`] = `
<ModalContent
  className="sto-modal-content"
>
  <AnimatedContentHeader
    activeClassName="sto-animated-header--active"
    className="sto-animated-header"
  >
    <Icon
      className="sto-animated-header-icon blurry"
      color="#95999D"
      height="64"
      icon="test-file-stub"
      spin={false}
      width="48"
    />
    <Icon
      className="sto-animated-header-exchange"
      color="#95999D"
      icon="exchange"
      spin={false}
    />
    <Icon
      className="sto-animated-header-icon"
      height="56"
      icon="test-file-stub"
      spin={false}
      width="56"
    />
  </AnimatedContentHeader>
  <Unknown
    className="sto-modal-title"
    ellipsis={false}
    tag="div"
  >
    Data access requirements
  </Unknown>
  <withRouter(Wrapper)
    app={
      Object {
        "editor": "Cozy",
        "manifest": Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/3.0.0/icon.svg",
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "permissions": Object {
            "albums": Object {
              "description": "Required to manage photos albums",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.photos.albums",
            },
            "apps": Object {
              "description": "Required by the cozy-bar to display the icons of the apps",
              "type": "io.cozy.apps",
              "verbs": Array [
                "GET",
              ],
            },
            "contacts": Object {
              "description": "Required to to share photos with your contacts",
              "methods": Array [
                "GET",
                "POST",
              ],
              "type": "io.cozy.contacts",
            },
            "files": Object {
              "description": "Required for photo access",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.files",
            },
            "settings": Object {
              "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
              "type": "io.cozy.settings",
              "verbs": Array [
                "GET",
              ],
            },
          },
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        },
        "slug": "photos",
        "type": "webapp",
        "version": "3.0.0",
      }
    }
  />
  <p
    className="u-error"
  >
    Something went wrong when fetching the application informations. Reason: Fetch mock error
  </p>
</ModalContent>
`;

exports[`InstallModalContent modal component should have install button busy if isInstalling 1`] = `
<ModalContent
  className="sto-modal-content"
>
  <AnimatedContentHeader
    activeClassName="sto-animated-header--active"
    className="sto-animated-header"
  >
    <Icon
      className="sto-animated-header-icon blurry"
      color="#95999D"
      height="64"
      icon="test-file-stub"
      spin={false}
      width="48"
    />
    <Icon
      className="sto-animated-header-exchange"
      color="#95999D"
      icon="exchange"
      spin={false}
    />
    <Icon
      className="sto-animated-header-icon"
      height="56"
      icon="test-file-stub"
      spin={false}
      width="56"
    />
  </AnimatedContentHeader>
  <Unknown
    className="sto-modal-title"
    ellipsis={false}
    tag="div"
  >
    Data access requirements
  </Unknown>
  <withRouter(Wrapper)
    app={
      Object {
        "editor": "Cozy",
        "manifest": Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/3.0.0/icon.svg",
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "permissions": Object {
            "albums": Object {
              "description": "Required to manage photos albums",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.photos.albums",
            },
            "apps": Object {
              "description": "Required by the cozy-bar to display the icons of the apps",
              "type": "io.cozy.apps",
              "verbs": Array [
                "GET",
              ],
            },
            "contacts": Object {
              "description": "Required to to share photos with your contacts",
              "methods": Array [
                "GET",
                "POST",
              ],
              "type": "io.cozy.contacts",
            },
            "files": Object {
              "description": "Required for photo access",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.files",
            },
            "settings": Object {
              "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
              "type": "io.cozy.settings",
              "verbs": Array [
                "GET",
              ],
            },
          },
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        },
        "slug": "photos",
        "type": "webapp",
        "version": "3.0.0",
      }
    }
  />
</ModalContent>
`;

exports[`InstallModalContent modal component should not display spinner if isFetching but during a cancel (isCanceling) 1`] = `
<ModalContent
  className="sto-modal-content"
>
  <AnimatedContentHeader
    activeClassName="sto-animated-header--active"
    className="sto-animated-header"
  >
    <Icon
      className="sto-animated-header-icon blurry"
      color="#95999D"
      height="64"
      icon="test-file-stub"
      spin={false}
      width="48"
    />
    <Icon
      className="sto-animated-header-exchange"
      color="#95999D"
      icon="exchange"
      spin={false}
    />
    <Icon
      className="sto-animated-header-icon"
      height="56"
      icon="test-file-stub"
      spin={false}
      width="56"
    />
  </AnimatedContentHeader>
  <Unknown
    className="sto-modal-title"
    ellipsis={false}
    tag="div"
  >
    Data access requirements
  </Unknown>
  <withRouter(Wrapper)
    app={
      Object {
        "editor": "Cozy",
        "manifest": Object {
          "categories": Array [
            "cozy",
          ],
          "developer": Object {
            "name": "Cozy",
            "url": "https://cozy.io",
          },
          "editor": "Cozy",
          "icon": "https://mockcozy.cc/registry/photos/3.0.0/icon.svg",
          "langs": Array [
            "en",
            "fr",
          ],
          "licence": "AGPL-3.0",
          "locales": Object {
            "en": Object {
              "changes": "### NEW!!!

 Now on the cozy-store :tada",
              "long_description": "### Features

Enjoy features to grab your data in you Cozy :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
            "fr": Object {
              "long_description": "### Photos

Gestionnaire de photos pour Cozy v3 :tada:",
              "short_description": "Gestionnaire de photos pour Cozy v3",
            },
          },
          "name": "Photos",
          "name_prefix": "Cozy",
          "permissions": Object {
            "albums": Object {
              "description": "Required to manage photos albums",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.photos.albums",
            },
            "apps": Object {
              "description": "Required by the cozy-bar to display the icons of the apps",
              "type": "io.cozy.apps",
              "verbs": Array [
                "GET",
              ],
            },
            "contacts": Object {
              "description": "Required to to share photos with your contacts",
              "methods": Array [
                "GET",
                "POST",
              ],
              "type": "io.cozy.contacts",
            },
            "files": Object {
              "description": "Required for photo access",
              "methods": Array [
                "GET",
                "POST",
                "PUT",
              ],
              "type": "io.cozy.files",
            },
            "settings": Object {
              "description": "Required by the cozy-bar to display Claudy and know which applications are coming soon",
              "type": "io.cozy.settings",
              "verbs": Array [
                "GET",
              ],
            },
          },
          "platforms": Array [
            Object {
              "type": "ios",
              "url": "",
            },
            Object {
              "type": "linux",
              "url": "",
            },
            Object {
              "type": "android",
              "url": "https://mock.app",
            },
          ],
          "routes": Object {
            "/": Object {
              "folder": "/",
              "index": "index.html",
              "public": false,
            },
            "/public": Object {
              "folder": "/public",
              "index": "index.html",
              "public": true,
            },
          },
          "short_description": "Photos manager for Cozy v3",
          "slug": "photos",
          "source": "registry://photos/stable",
          "type": "webapp",
          "version": "3.0.0",
        },
        "slug": "photos",
        "type": "webapp",
        "version": "3.0.0",
      }
    }
  />
</ModalContent>
`;

exports[`InstallModalContent modal component should return a spinner if isFetching 1`] = `
<ModalContent
  className="sto-modal-content"
>
  <div
    className="sto-install-loading"
  >
    <Wrapper
      size="xlarge"
    />
  </div>
</ModalContent>
`;
